#57 Fix binary blob handling with MariaDB Connector/J 3.5.2+

solita-ville-autio opened the pull request:
Pull request for #57 as requested (pun intended).

After some initial misadventures with different exploratory approaches it turned out that the suggested solution of passing down the dialect to `ResultSetUtils` and just overriding whatever types `ResultSetMetaData` returns didn't require that extensive changes after all.

There are obviously multiple ways how to handle the driver version specific stuff and I opted to use this rudimentary version parser found from some gist that doesn't understand more exotic version strings but anything more sophisticated stuff seemed like overkill for this case, and I didn't want to introduce another runtime dependency with some SemVer parser library.

The changes were tested with MariaDB Connector/J versions 2.7.12, 3.5.1 and 3.5.3 with both basic containerized tests and our actual application tests; no issues were encountered.

solita-ville-autio committed:
#57 Fixed binary blob handling with MariaDB Connector/J 3.5.2+

Root cause is that the driver returns java.sql.Blob as the column class
name in ResultSetMetaData while returning the contents directly as byte
arrays when fetched.

Since this behavior is so database and driver specific and causes issues
in instantiator lookup, minimally interfering solution is to use dialect
to provide type overrides to ResultSetUtils to make sure that the
expected and actually received types match.

komu commented:
I haven't yet had a proper look at the actual code, but it seems that there are quite a lot of unrelated changes in this PR (build files etc). Could you please remove those?

solita-ville-autio commented:
Yeah sure, just a bad habit of mine to always touch that stuff.

komu commented on dalesbred/src/main/java/org/dalesbred/dialect/Dialect.java:
```diff
@@ -53,6 +55,8 @@ public abstract class Dialect {
 
     private static final Logger log = LoggerFactory.getLogger(Dialect.class);
 
+    protected final Map<String, Class<?>> resultSetMetaDataTypeOverrides = new ConcurrentHashMap<>();
```
Since there a getter for this anyway, I think it's clearer to leave this private. Or if the getter always returns an immutable map, then document that somehow. Anyway, the contract given to subclasses should be clear.

komu commented on dalesbred/src/main/java/org/dalesbred/dialect/Dialect.java:
```diff
@@ -159,4 +163,41 @@ public void registerTypeConversions(@NotNull TypeConversionRegistry typeConversi
         ArgumentBinder.bindArgument(ps, index, value);
     }
+
+    public Map<String, Class<?>> getResultSetMetaDataTypeOverrides() {
```
Dalesbred uses explicit nullability annotations everywhere, please add them here as well.

Should this return an immutable map or are the callers allowed to change the returned map? Either way, I guess the doc could mention that.

_Update: we probably don't want this method at all, see later comments._

komu commented on dalesbred/src/main/java/org/dalesbred/dialect/Dialect.java:
```diff
@@ -159,4 +163,41 @@ public void registerTypeConversions(@NotNull TypeConversionRegistry typeConversi
+
+        protected static Version withNumber(@NotNull String version) {
+            return new Version(version);
+        }
```
This throws for unknown version patterns, which is bad, I think. At the least, it should be documented, since this is visible in the public API. And if this throws, then I guess the MariaDBDialect could catch the error and log it. It's not nice of the autodetection kludge prevents the code from working at all.

That said, perhaps this class could be moved to `internal`-package. It's ok that the implementation of MariaDBDialect uses the class, but I don't think we want to increase the API surface for this. If someone wants to parse versions, they'd better of using their own solution instead of something that we support.

Also, perhaps `parse` would be a better name?

komu commented on dalesbred/src/main/java/org/dalesbred/dialect/Dialect.java:
```diff
@@ -159,4 +163,41 @@ public void registerTypeConversions(@NotNull TypeConversionRegistry typeConversi
+                    return left < right ? -1 : 1;
+            }
+            return 0;
+        }
```
The style used elsewhere does not use `final` for local variables. I don't really feel strongly either way, but it's good to be consistent.

For members, I use final all the time, but for locals where the scope is small anyway, I don't really want it.

I think a more natural implementation would walk both arrays to their minLength and then return -1 or 1 if one of the arrays had more elements after that. Of course Arrays.compare would do just that, but unfortunately that uses Java 9 and I don't really want to bump the Java version up just for that. Anyway, perhaps just:

```java
@Override
public int compareTo(@NotNull Version another) {
    for (int i = 0; i < min(numbers.length, another.numbers.length); i++) { 
        int result = Integer.compare(numbers[i], another.numbers[i]);
        if (result != 0)
            return result;
    }

    return Integer.compare(numbers.length, another.numbers.length);
}
```

komu commented on dalesbred/src/main/java/org/dalesbred/dialect/Dialect.java:
```diff
@@ -159,4 +163,41 @@ public void registerTypeConversions(@NotNull TypeConversionRegistry typeConversi
+
+        protected boolean isGreaterThanOrEqualTo(Version another) {
+            return this.compareTo(another) >= 0;
+        }
```
I think this is superfluous. A Java-programmer should see `a.compareTo(b) >= 0` and see `a >= b` in their mind. Of course it's ugly and it's nice that Kotlin does that automatically, but still I think the compareTo-idiom is so common that there's no need to add method just to hide that.

solita-ville-autio replied:
Yeah probably, this is just one of those things that I personally cannot for the life of me ever remember which way it goes. 😅

komu replied:
Well, it's pretty logical once you learn the rule: you keep the operands in order and just replace `compareTo` with the operator:

* `a.compareTo(b) >= 0` -> `a >= b`
* `a.compareTo(b) < 0` -> `a < b`

But sure, it's annoying noise.

komu commented on dalesbred/src/main/java/org/dalesbred/internal/instantiation/SqlArrayConversion.java:
```diff
@@ -37,17 +37,17 @@
+    private SqlArrayConversion(@NotNull Type elementType, @NotNull InstantiatorProvider instantiatorRegistry) {
         this.elementType = elementType;
-        this.instantiatorRegistry = instantiatorRegistry;
+        this.instantiatorRegistry = requireNonNull(instantiatorRegistry);
```
I haven't bothered with `requireNonNull` in cases where the API is internal and there is a `@NotNull` annotation in place. Anyway, if it's added for one parameter, why not the other?

So I guess this change is left over from some work and is not really needed.

komu commented on dalesbred/src/main/java/org/dalesbred/internal/jdbc/ResultSetUtils.java:
```diff
@@ -37,20 +38,23 @@ public final class ResultSetUtils {
         String className = metaData.getColumnClassName(column);
 
+        if (dialect.getResultSetMetaDataTypeOverrides().containsKey(className))
+            return dialect.getResultSetMetaDataTypeOverrides().get(className);
```
The `containsKey` + `get` -pattern does two map calls for no good reason. Basically the only reason for doing `containsKey` with `get` is if your map might contain null elements and you got a null back from `get`. In that case you can make another check if the value was in the map. But here that's not needed and you can say just:

```java
Type type = dialect.getResultSetMetaDataTypeOverrides().get(className);
if (type != null)
    return type
```

That said, exposing the whole map outside seems a bit ugly when in reality you just want the dialect to be able to override the type for override. So perhaps the dialet should have:

```java
public @Nullable Type overrideResultSetMetaDataType(@NotNull String className) {
     return null;
}
```

Then the MariaDBDialect could implement this as it wants and all of my comments about map-mutability etc. in the beginning go away.

komu commented on dalesbred/src/main/java/org/dalesbred/internal/jdbc/ResultSetUtils.java:
```diff
@@ -37,20 +38,23 @@ public final class ResultSetUtils {
+
         // MariaDB Connector/J 3.x encodes byte array types in a way that is incompatible with Class.forName
         if (className.equals("byte[]"))
             return byte[].class;
```
Perhaps this could be moved to the MariaDBDialect as well.

solita-ville-autio replied:
I thought of that too but since it's theoretically possible that this would occur with some other driver too I decided to leave it here. But since we're now going down this path of driver specific workarounds in dialects, it does make sense to move this kind of kludge there.

komu reviewed (changes requested):
Thanks! Overall seems good.

I had some comments about the API surface. Basically, we only want one new method visible in non-internal packages, `overrideResultSetMetaDataType` in the `Dialect`.

solita-ville-autio commented:
Thanks for the comprehensive review! I'll try to get back to you sometime next week with the requested changes.

solita-ville-autio committed:
#57 Changes based on review comments

- Reduced public API footprint
- Version class moved to its own file in internal.utils package
- If driver version number parsing fails for some reason, the exception
is now caught and logged instead of just failing
- Moved previous MariaDB Connector/J 3.x workaround to dialect
- Various small code style changes

komu commented on dalesbred/src/main/java/org/dalesbred/dialect/MariaDBDialect.java:
```diff
@@ -22,20 +22,44 @@
 
+    private static final Logger log = LoggerFactory.getLogger(MariaDBDialect.class);
+
+    private final @NotNull Map<String, Type> resultSetMetaDataTypeOverrides = new ConcurrentHashMap<>();
```
This is only written in the constructor and effectively immutable after that, so a normal unsynchronized HashMap would provide better concurrency.

komu commented on dalesbred/src/main/java/org/dalesbred/dialect/MariaDBDialect.java:
```diff
@@ -22,8 +22,44 @@
+
+    private final @NotNull Map<String, Type> resultSetMetaDataTypeOverrides = new ConcurrentHashMap<>();
+
+    public MariaDBDialect(@NotNull String driverVersion) {
```
I just realized that this is a breaking change since there's no default constructor anymore. If someone was instantiating the dialect manually, it won't work anymore. So I guess there are three ways to go:

1. accept the breaking change since it's such a small change anyway
2. keep this constructor but add a default constructor that uses the old semantics
3. move the flag detection away from the constructor

Perhaps something like the following:

```java
public class MariaDBDialect extends Dialect {
    private boolean blobsAsByteArrays = false;

    public void setBlobsAsByteArrays(boolean blobsAsByteArrays) {
        this.blobsAsByteArrays = blobsAsByteArrays;
    }

    public void autodetectSettings(metadata: DatabaseMetaData) {
        Version version = Version.parse(metadata. getDriverVersion()));
        if (version == null)
            log.error("Could not parse driver version '{}'", metadata. getDriverVersion());

        blobsAsByteArrays  = version.compareTo(new Version(3, 5, 2)) >= 0;
    }

   @Override
    public @Nullable Type overrideResultSetMetaDataType(@NotNull String className) {
        switch (className) {
        case "byte[]": 
            // MariaDB Connector/J 3.x encodes byte array types in a way that is incompatible with Class.forName
            return byte[].class;

        case "java.sql.Blob":
            // MariaDB Connector/J 3.5.2+ reports java.sql.Blob as the class name for
            // binary blobs but returns them directly as byte arrays when fetched  
            return blobsAsByteArrays ? byte[].class : null;
        default:
            return null;
        }
    }
}
```

This way the autodetection code could call the default constructor followed by a call to `autodetectSettings`, users giving the dialect explicitly could call `setBlobsAsByteArrays(true)` and nothing would break for existing users.

Maintaining API compatibility in a library is easy to forget. Normally you could just refactor and fix all the callers, but there's no option for that here. That's why I really want to publish as little as possible (e.g. move `Version` to the internal package).

solita-ville-autio replied:
Lack of default constructor came to my mind also last week and I thought that maybe I could include one that just makes the bold assumption that the driver version is at least 3.5.2 but I figured that it wouldn't probably pass your review and in any case since there was only one place where the dialect was instantiated this wouldn't be an issue. It didn't occur to me that someone might want to instantiate the dialect manually elsewhere.

This is just nitpicking but technically speaking MariaDBDialect is not included in the latest officially released version so any breaking changes to it just affect people who are maintaining their own private forks that were cloned after last May.

But I digress, I think these changes are still reasonable and hopefully we get rid of that Map once and for all.

komu replied:
Oh, I didn't realize that MariaDBDialect was not already released. 😆 I make no guarantee to keep unreleased APIs stable. Sorry about that, I bothered you for no reason.

solita-ville-autio committed:
#57 Use default constructor w/ separate method for feature detection

solita-ville-autio commented:
This just now pushed version is probably the cleanest that I could within reason come up with based on your suggestion. This had a potential to turn into a real quagmire due to `DatabaseMetaData` declaring that `getDriverVersion()` throws a `SQLException` when combined with the maybe a bit too pedantic desire to protect future generations from shooting themselves in the foot with this ad hoc version number parser. I ended up using a checked exception since constructors + unchecked exceptions + potential nullability just seemed to turn this into a real mess (at least if you want IDEA to be happy and not complain about pointless null checks). I'd rather not spend too much more time in quest for pure perfection.

komu reviewed (approved):
Thanks for your work and for coping with my requests for changes!

I'll try to get this released soon. I'd make the release right away, but the method of deploying changes to Maven Central was discontinued this summer, and I'll have to migrate to the new deployment method.
